src/charm.py

```diff
@@ -105,228 +110,145 @@ def _on_pgbouncer_pebble_ready(self, event: PebbleReadyEvent) -> None:
                 in the container. When config files are updated, pgbouncer must be restarted for
                 the changes to take effect. However, these config updates can be done in batches,
                 minimising the amount of necessary restarts.
         """
-        pgb_container = self.unit.get_container(self._pgbouncer_container)
-        if users is None:
-            users = self._get_userlist_from_container()
-        userlist = self._generate_userlist(users)
-
-        try:
-            # Check that we're not updating this file unnecessarily
-            if pgb_container.pull(USERLIST_PATH).read() == userlist:
-                logger.info("updated userlist does not modify existing pgbouncer userlist")
-                return
-        except FileNotFoundError:
-            # there is no existing userlist.txt file, so carry on and add one.
-            pass
+        pgb_container = self.unit.get_container(PGB)
+        if not pgb_container.can_connect():
+            self.unit.status = WaitingStatus(
+                "Unable to push config to container - container unavailable."
+            )
+            return
 
         pgb_container.push(
             USERLIST_PATH,
-            userlist,
-            user=self._pgbouncer_user,
-            permissions=0o600,
+            pgb.generate_userlist(userlist),
+            user=PG_USER,
+            permissions=0o777,
             make_dirs=True,
         )
         logging.info("pushed new userlist to pgbouncer container")
 
         if reload_pgbouncer:
             self._reload_pgbouncer()
 
-    def _generate_userlist(self, users: Dict[str, str]) -> str:
-        """Generate userlist.txt from the given dictionary of usernames:passwords.
-
-        Args:
-            users: a dictionary of usernames and passwords
-        Returns:
-            A multiline string, containing each pair of usernames and passwords separated by a
-            space, one pair per line.
-        """
-        return "\n".join([f'"{username}" "{password}"' for username, password in users.items()])
-
-    def _get_users_from_charm_config(self, users: Dict[str, str] = {}) -> Dict[str, str]:
-        """Imports users from charm config and generates passwords when necessary.
-
-        When generated, passwords are hashed using md5.
-
-        Args:
-            users: an existing dictionary of usernames and passwords
-        Returns:
-            the `users` dictionary, with usernames from charm config and corresponding generated
-            passwords appended.
-        """
-        for admin_user in self.config["pgb_admin_users"].split(","):
-            if admin_user not in users or users[admin_user] is None:
-                users[admin_user] = self._hash(self._generate_password())
-        return users
-
-    def _get_userlist_from_container(self) -> Dict[str, str]:
+    def _read_userlist(self) -> Dict[str, str]:
```

Comment:
This function seems not to be used (it's used only in the unit tests).

Reply:
removed